src/main/java/ca/utoronto/fitbook/adapter/persistence/firebase/ExerciseFirebaseRepository.java

```diff
@@ -0,0 +1,63 @@
+package ca.utoronto.fitbook.adapter.persistence.firebase;
+
+import ca.utoronto.fitbook.adapter.persistence.GenericRepository;
+import ca.utoronto.fitbook.entity.Exercise;
+import ca.utoronto.fitbook.entity.RepetitiveExercise;
+import ca.utoronto.fitbook.entity.TemporalExercise;
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.DocumentSnapshot;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Repository;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+
+@Repository
+public class ExerciseFirebaseRepository implements GenericRepository<Exercise>
+{
+    private final FirebaseDatastore datastore;
+
+    public ExerciseFirebaseRepository() {
+        this.datastore = FirebaseDatastore.getInstance();
+    }
+
+    @ResponseStatus(value=HttpStatus.INTERNAL_SERVER_ERROR, reason="Invalid exercise type")
+    private static class InvalidExerciseTypeException extends RuntimeException {
+        public InvalidExerciseTypeException(String type) {
+            super(String.format("Exercise type (%s) is invalid.", type));
+        }
+    }
+
+    /**
+     * @param id Id of the exercise
+     * @return the exercise with the given Id
+     */
+    @Override
+    public Exercise getById(String id) {
+        ApiFuture<DocumentSnapshot> future = datastore.getCollection("exercises").document(id).get();
+        try {
+            DocumentSnapshot document = future.get();
+            if (document.exists()) {
+                String type = document.getString("type");
+                if (Objects.equals(type, Exercise.ExerciseType.REPETITIVE.toString()))
+                    return document.toObject(RepetitiveExercise.class);
+                else if (Objects.equals(type, Exercise.ExerciseType.TEMPORAL.toString()))
+                    return document.toObject(TemporalExercise.class);
```

Comment:
I believe that this section of code in the ExerciseFirebaseRepository may be violating the Open/Closed SOLID principle, as any extension for a new kind of exercise would require the direct modification of the repository and adding new conditions.

Reply:
Sounds good, will push a change.